### ISSUE TITLE
Better playlist editing

### ISSUE THREAD

**Issue by wm4** [cmdrkotori/mpc-qt#247]
I'd love to have more convenient ways to edit the playlist. I suggest the following extensions or improvements:

1. When copy&pasting an entry, the pasted entry is currently always appended to the end of the list. Also it flickers for a moment and something strange happens to the keyboard focus. I'd expect it to be inserted before the current selection, and after inserting, the first entry should be selected (or maybe all inserted entries should be selected). Not sure how the user could achieve inserting them at the end of the list.
2. Move selection (one or multiple entries) up and down the list if Ctrl is held down while pressing the up/down keys. For example if you have 3 entries selected, and then press Ctrl+Down, the first entry after the selection is moved to before the first selected entry.
3. Make it react to standard copy paste key shortcuts (Ctrl+V/C/X). Currently, e.g. copy selection is on Ctrl+Shift+C instead of Ctrl+C. Not sure why (there's probably a good reason behind that), but enabling the standard keys would be a worthy goal.
4. There is no way to insert new files at the current position in the playlist. "Quick Add To Playlist" appends it at the end. Maybe it should insert at the current selection. (Like in 1., I'm not sure how the user could append it at the end if that is implemented.)
5. The playlist tabs have a close button. If you press it, the playlist is apparently gone forever. Not good, it's too easy to delete them. Maybe playlists could be stored as separate files on disk, which aren't deleted when the playlist is closed? Or playlists could be kept in a separate list, with the tabs only showing "visible" playlists. To delete a playlist finally, the user would have to press some kind of delete button in that list.
6. When you add a new playlist, you're prompted for a name. And then there's apparently no way to rename it. It should be the other way around: a playlist is added without prompting the user for a name (instead use a default name like "Playlist 3"), and provide a way to rename the playlist any time.
7. Currently, playlists are saved to disk only when mpc-qt is closed. This makes me close and restart the player every time I edit the playlist. Otherwise there's the danger that the change gets lost when mpc-qt gets killed. (mpc-qt has an uptime of weeks and even months for me, but sometimes it goes violently.)
8. Maybe there should be an undo function for playlist edits, or a list of backups of previous versions. When I mess something up when editing the playlist, I sometimes kill -9 mpc-qt to restore the previous version. Backups may also be a solution to implement entry 5 in this list.

I can split this into multiple issues if wanted. Maybe 1 for each item in the list above.

Not sure if these requests are reasonable.

**Comment by cmdrkotori** [comment]
1. Yes this should be customizable for sure.

2. Possible.

3. I agree with you, standard stuff would be nicer. However I have to stick with the home cinema defaults due to the principle of least surprise. The conflict in question is principally
`[34.875120631] [action] "Paste" conflicts with "Open De&vice..." because of key sequence "Ctrl+V". "Open De&vice..."'s key sequence was removed.` So the rest are shift'd to be consistent. Some terminals use the shift'd sequences anyway.

4. If it's implemented drag & drop should still work.

5. Very good point. Hiding playlists and soft-deleting would be a very nice feature to have. (also good for users hiding the obvious).

6. Double click the playlist tab. But it should probably be in the context menu too at least.

7. I did this to prevent SSD thrashing. Yeah saving at exit would be a problem if it's left open for weeks at a time. Perhaps an instant write mode could be added and a choice like [At exit/When editing/Always] and maybe depending on implementation a checkbox [Treat playback progression the same as an edit].

8. More or less the same feature as 5.

Overall good points about the playlist persistence and gotchas.

**Comment by wm4** [comment]
> I did this to prevent SSD thrashing. Yeah saving at exit would be a problem if it's left open for weeks at a time.

A solution might be to write changes back only after 1 minute of inactivity or so. Then it won't do many writes for many small changes.

> [34.875120631] [action] "Paste" conflicts with "Open De&vice..." because of key sequence "Ctrl+V". "Open De&vice..."'s key sequence was removed.

That seems awfully obscure, might be worth changing the bindings for that one.

Anyway, thanks for considering implementing these.
